spec/controllers/users/phone_confirmation_controller.rb

```diff
@@ -0,0 +1,106 @@
+
+describe Users::PhoneConfirmationController, devise: true do
+  describe 'before_actions' do
+    it 'includes authentication checks' do
+      expect(subject).to have_filters(
+        :before,
+        :authenticate_user!,
+        :check_for_unconfirmed_mobile
+      )
+    end
+  end
+
+  describe '#new' do
+    before do
+      sign_in_as_user
+    end
+
+    context 'when :unconfirmed_mobile is not set in session' do
+      it 'redirects to rool_url' do
+        get :new
+
+        expect(response).to redirect_to(root_url)
+      end
+    end
+
+    context 'when :unconfirmed_mobile is set in session' do
+      before do
+        subject.user_session[:unconfirmed_mobile] = '+1 (555) 555-5555'
+      end
+
+      it 'generates a new confirmation code in the session' do
+        expect(subject.user_session[:phone_confirmation_code]).to be_nil
+        get :new
+        expect(subject.user_session[:phone_confirmation_code]).not_to be_nil
+      end
+
+      it 'sends new confirmation via SMS' do
+        allow(SmsSenderOtpJob).to receive(:perform_later)
+        get :new
+
+        expect(SmsSenderOtpJob).to have_received(:perform_later).
+          with(subject.user_session[:phone_confirmation_code],
+               subject.user_session[:unconfirmed_mobile])
+      end
+    end
+  end
+
+  describe '#confirm' do
+    before do
+      sign_in_as_user
+      subject.user_session[:unconfirmed_mobile] = '+1 (555) 555-5555'
+      subject.user_session[:phone_confirmation_code] = '123'
+    end
+
+    context 'when given valid code' do
+      before do
+        post :confirm, code: '123'
+      end
+
+      it 'clears session data' do
+        expect(subject.user_session[:unconfirmed_mobile]).to be_nil
+        expect(subject.user_session[:phone_confirmation_code]).to be_nil
+      end
+
+      it 'redirects to edit_user_registration_path' do
+        expect(response).to redirect_to(edit_user_registration_path)
+      end
+
+      it 'displays success flash notice' do
+        expect(flash[:success]).to eq t('upaya.notices.phone_confirmation_successful')
+      end
+    end
+
+    context 'when given invalid code' do
+      before do
+        post :confirm, code: '999'
+      end
+
+      it 'does not clear session data' do
+        expect(subject.user_session[:unconfirmed_mobile]).to eq('+1 (555) 555-5555')
+        expect(subject.user_session[:phone_confirmation_code]).to eq('123')
+      end
+
+      it 'redirects back phone_confirmation_path' do
+        expect(response).to redirect_to(phone_confirmation_path)
+      end
+
+      it 'displays error flash notice' do
+        expect(flash[:error]).to eq t('errors.messages.invalid_confirmation_code')
+      end
+    end
+  end
+
+  describe '#show' do
+    before do
+      sign_in_as_user
+      subject.user_session[:unconfirmed_mobile] = '+1 (555) 555-5555'
+      get :show
+    end
+
+    it 'renders :show templatesent' do
+      expect(response).to render_template(:show)
+    end
+  end
```

Comment:
Please add a test to verify that the `@code_value` variable only gets set when prefill_otp_codes is true, and then add a  `spec/lib/feature_management_spec` unit test for the FeatureManagement method to makes sure it only returns true when in development and sms is disabled.